frontend/packages/console-plugin-sdk/src/typings/index.ts

```diff
@@ -0,0 +1,66 @@
+/**
+ * An extension of the Console web application.
+ *
+ * Each extension is a realization (instance) of an extension `type` using the
+ * parameters provided via the `properties` object.
+ *
+ * Core extension types should follow `Category` or `Category/Specialization`
+ * format, e.g. `NavItem/Href`.
+ *
+ * @todo(vojtech) write ESLint rule to guard against extension type duplicity
+ */
+export interface Extension<P> {
```

Comment:
Can we use `type` here to discourage using classes?

Reply:
In TypeScript, [type aliases](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#3.10) provide a name to refer to any kind of type, while [interfaces](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#7) are used to name object specific types.

> Interface types have many similarities to type aliases for object type literals, but since interface types offer more capabilities they are generally preferred to type aliases.

> An interface can be named in an extends or implements clause, but a type alias for an object type literal cannot.

> An interface can have multiple merged declarations, but a type alias for an object type literal cannot.

When describing object types, I'd stick to the TS spec recommendation and prefer using interfaces over type aliases. Using interfaces does **not** imply using classes to implement those interfaces.

Reply:
> An interface can be named in an extends or implements clause, but a type alias for an object type literal cannot.

Not true. You can still implement a type; unless the type definition uses a union operator.

Reply:
```ts
type TestObj = {
  x: number;
  y: () => boolean;
};

// valid
class Foo implements TestObj {
  x = 1;
  y() { return true; }
}
```

@christianvogt You are correct - you can use object literal type alias (e.g. `TestObj`) in the `implements` clause. It's also possible to do object composition via the intersection (`&`) operator:

```ts
type AnotherTestObj = TestObj & {
  z: string;
};
```

However, the spec is still correct in that you can't use e.g. `TestObj` in the `extends` clause, more specifically when declaring a class:

```ts
// valid
type Bar = <T extends TestObj>(obj: T) => void;

// invalid
class Qux extends TestObj {}
```

Still, I'd prefer using interfaces, as they describe the shape of an object (they're not a traditional OOP interface whose usage is coupled with classes).